gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/computer/GraphFilter.java

```diff
@@ -0,0 +1,193 @@
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+/**
+ * GraphFilter is used by {@link GraphComputer} implementations to prune the source graph data being loaded into the OLAP system.
+ * There are two types of filters: a {@link Vertex} filter and an {@link Edge} filter.
+ * The vertex filter is a {@link Traversal} that can only check the id, label, and properties of the vertex.
+ * The edge filter is a {@link Traversal} that starts at the vertex are emits all legal incident edges.
+ * If no vertex filter is provided, then no vertices are filtered. If no edge filter is provided, then no edges are filtered.
+ * The use of a GraphFilter can greatly reduce the amount of data processed by the {@link GraphComputer}.
+ * For instance, for {@code g.V().count()}, there is no reason to load edges, and thus, the edge filter can be {@code bothE().limit(0)}.
+ *
+ * @author Marko A. Rodriguez (http://markorodriguez.com)
+ */
+public final class GraphFilter implements Cloneable, Serializable {
+
+    public enum Legal {
+        YES, NO, MAYBE;
+
+        public boolean positive() {
+            return this != NO;
+        }
+
+        public boolean negative() {
+            return this == NO;
+        }
+    }
+
+    private Traversal.Admin<Vertex, Vertex> vertexFilter = null;
+    private Traversal.Admin<Vertex, Edge> edgeFilter = null;
+
+    private boolean allowNoEdges = false;
+    private Direction allowedEdgeDirection = Direction.BOTH;
+    private Set<String> allowedEdgeLabels = new HashSet<>();
+    //private boolean allowAllRemainingEdges = false;
+
+    public void setVertexFilter(final Traversal<Vertex, Vertex> vertexFilter) {
+        if (!TraversalHelper.isLocalVertex(vertexFilter.asAdmin()))
+            throw GraphComputer.Exceptions.vertexFilterAccessesIncidentEdges(vertexFilter);
+        this.vertexFilter = vertexFilter.asAdmin().clone();
+    }
+
+    public void setEdgeFilter(final Traversal<Vertex, Edge> edgeFilter) {
+        if (!TraversalHelper.isLocalStarGraph(edgeFilter.asAdmin()))
+            throw GraphComputer.Exceptions.edgeFilterAccessesAdjacentVertices(edgeFilter);
+        this.edgeFilter = edgeFilter.asAdmin().clone();
+        if (this.edgeFilter.getEndStep() instanceof RangeGlobalStep && 0 == ((RangeGlobalStep) this.edgeFilter.getEndStep()).getHighRange())
+            this.allowNoEdges = true;
+        else if (this.edgeFilter.getStartStep() instanceof VertexStep) {
+            this.allowedEdgeLabels.clear();
+            this.allowedEdgeLabels.addAll(Arrays.asList(((VertexStep) this.edgeFilter.getStartStep()).getEdgeLabels()));
+            this.allowedEdgeDirection = ((VertexStep) this.edgeFilter.getStartStep()).getDirection();
+            //this.allowAllRemainingEdges = 1 == this.edgeFilter.getSteps().size();
+        }
+    }
+
+    /*public void compileFilters() {
```

Comment:
I was using `compileFilters()` at first, but then didn't need them. I left it commented out just in case its something that might pop up in the future. `GraphFilter` is very new and over this 3.2.0 push, I'm sure we will tweak here and there. Just want people to know its around.

Regarding the `allowRemainingEdges`. Again, its something I didn't need, but can easily add if we find a good use case.